src/Compilers/Core/Portable/MetadataReference/MetadataReference.cs

```diff
@@ -155,7 +155,12 @@ public MetadataReference WithProperties(MetadataReferenceProperties properties)
             DocumentationProvider? documentation = null,
             string? filePath = null)
         {
-            var metadata = AssemblyMetadata.CreateFromImage(peImage);
+            Metadata metadata = properties.Kind switch
```

Comment:
>Metadata metadata = properties.Kind switch [](http://example.com/codeflow?start=12&length=42)

I think image knows whether it is a module or an assembly. This cannot be overridden by a parameter. It feels like we should check how compiler figures out what references it gets and use the same logic here.

Reply:
Looking at implementation of `AssemblyMetadata.CreateFromImage`, it doesn't matter. The `peImage` is always parsed as a module, then wrapped in `AssemblyMetadata` if needed. So I guess the parameter is caller's way of telling which instance they want (`ModuleMetadata` or `AssemblyMetadata`).

Reply:
> we should check how compiler figures out what references it gets

Seems that `csc.exe` uses `MetadataReference.CreateFromFile`:

https://github.com/dotnet/roslyn/blob/76cd2f1f0d55172a428e4fdae95fd7b0a2b2682f/src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs#L211

Reply:
Interesting. An assembly includes an assembly manifest, a module doesn't. It feels like there is a way to check for that. And it looks like we do check. For example, if I supply a module in place of an assembly, I get "error CS1509: The referenced file '....netmodule' is not an assembly", If an assembly is supplied in place of a module, "error CS1542: '....dll' cannot be added to this assembly because it already is an assembly" At what point do we perform the checks?

Reply:
> At what point do we perform the checks?

This is checked by reference manager - [method `ResolveMetadataReferences`](https://sourceroslyn.io/#Microsoft.CodeAnalysis/ReferenceManager/CommonReferenceManager.Resolution.cs,ba8c4b46d9a0d4f7). That's called after `MetadataReference`s are created. So the compiler creates `MetadataReference`s with the given type and in a separate step checks they are correct. Hence `MetadataReference.Create*` factory methods shouldn't perform these checks, right?